File: GenericCalculator.cpp
#include <iostream>
#include <stdexcept>

template<typename T>
class GenericCalculator {
public:
    void calculate(T x, T y, MathOperation<T>& mathOperation) {
        T result = mathOperation.perform(x, y);
        std::cout << "Result: " << result << std::endl;
    }

    class MathOperation {
    public:
        virtual T perform(T a, T b) = 0;
        virtual ~MathOperation() {}
    };

    class Addition : public MathOperation {
    public:
        T perform(T a, T b) override {
            return a + b;
        }
    };

    class Subtraction : public MathOperation {
    public:
        T perform(T a, T b) override {
            return a - b;
        }
    };

    class Multiplication : public MathOperation {
    public:
        T perform(T a, T b) override {
            return a * b;
        }
    };

    class Division : public MathOperation {
    public:
        T perform(T a, T b) override {
            if (b == 0) {
                throw std::invalid_argument("Division by zero is not allowed");
            }
            return a / b;
        }
    };
};

int main() {
    GenericCalculator<double> calculator;
    calculator.calculate(10.0, 2.0, calculator.Addition());
    calculator.calculate(10.0, 2.0, calculator.Subtraction());
    calculator.calculate(10.0, 2.0, calculator.Multiplication());
    calculator.calculate(10.0, 2.0, calculator.Division());

    return 0;
}

File: GenericCalculator.cs
using System;

public class GenericCalculator<T>
{
    public void Calculate(T x, T y, IMathOperation<T> mathOperation)
    {
        T result = mathOperation.Perform(x, y);
        Console.WriteLine("Result: " + result);
    }

    public interface IMathOperation<T>
    {
        T Perform(T a, T b);
    }

    public class Addition : IMathOperation<T>
    {
        public T Perform(T a, T b)
        {
            dynamic aa = a;
            dynamic bb = b;
            return aa + bb;
        }
    }

    public class Subtraction : IMathOperation<T>
    {
        public T Perform(T a, T b)
        {
            dynamic aa = a;
            dynamic bb = b;
            return aa - bb;
        }
    }

    public class Multiplication : IMathOperation<T>
    {
        public T Perform(T a, T b)
        {
            dynamic aa = a;
            dynamic bb = b;
            return aa * bb;
        }
    }

    public class Division : IMathOperation<T>
    {
        public T Perform(T a, T b)
        {
            dynamic aa = a;
            dynamic bb = b;
            if (bb == 0)
            {
                throw new ArgumentException("Division by zero is not allowed");
            }
            return aa / bb;
        }
    }
}

class Program
{
    static void Main(string[] args)
    {
        GenericCalculator<double> calculator = new GenericCalculator<double>();
        calculator.Calculate(10.0, 2.0, new GenericCalculator<double>.Addition());
        calculator.Calculate(10.0, 2.0, new GenericCalculator<double>.Subtraction());
        calculator.Calculate(10.0, 2.0, new GenericCalculator<double>.Multiplication());
        calculator.Calculate(10.0, 2.0, new GenericCalculator<double>.Division());
    }
}
